### DISCUSSION TITLE
Some keycloak users have multiple passwords

### DISCUSSION THREAD

**Discussion by arvin-mumba** [keycloak/keycloak#31556] [Q&A]
Hello everyone, I'd like to share a recent issue we are experiencing wherein some of our users has recently reported that they cannot log in after changing their passwords (via the Forgot Password feature).

We had a little bit of investigation and it turns out that the users that cannot login actually has multiple passwords linked to their account.

It's really puzzling as to how this happened, and we are clueless on how to reproduce this, and it's only happening to very few users. I've tried replicating it by going through the forgot password flow. But wasn't able to successfully reproduce the issue.

The keycloak realm config is as bare as it can be (I've also attached as screenshot of the Realm Settings -> Login config)
<img width="881" alt="realm setting" src="https://github.com/user-attachments/assets/4c1d9037-7f9c-4301-9c0f-44ec0f6949f0">
If it helps, we are currently on Keycloak v21 (recently upgraded from v14).

I am hoping someone else has already experienced this before and has also found the root cause and solution.

Thank you

**Comment by darius-m** [comment]
I have not personally seen this issue, but from the screenshot you have attached, the thing that stands out is Duplicate emails. Are there any duplicate emails in the accounts that have encountered this issue?

On a side note, it is strongly recommended that verify email is set. If I remember correctly, that setting enforces user confirmation of ownership of an email address before allowing them to update the email linked to the account. Not verifying the email could lead to impersonation issues in some cases.

**Reply by arvin-mumba** [reply]
Hi @darius-m, upon checking none of the users affected have duplicate emails (even for the non affected users, none of them have duplicate emails as well). I wonder what else could have possibly triggered the multiple passwords issue.

To add, there was someone who posted around 3 years ago, with a similar issue as ours. However, there's not much information from that one as well. https://keycloak.discourse.group/t/user-with-2-passwords/8415

I've also taken note about your recommendation for the verify email as well. Thank you

**Reply by darius-m** [reply]
Perhaps there was an issue with the password update process and the old password was not correctly matched for the user (and consequently was not updated) or the process failed to delete the old password for some reason? Does the issue happen reliably for a particular user, or is it random?

On a different note, Keycloak does not have support for older versions (except for the RedHat distribution, but even that I think is supported for version 22, so version 21 is not really supported in any notable way). It would be probably useful if you could upgrade to see if the issue has been fixed in the newest version; if it was not, maybe the Keycloak team can investigate a bit if you find a lead.

**Reply by arvin-mumba** [reply]
The issue happened to very few random users and was not happening reliably/consistently. After we removed the multiple passwords, the issue was no longer happening for those users. And as of now, no other users has reported the issue yet.

We'll try to observe for now if there will be new users who will experience the same issue. Hopefully we can get more information from those to aid in the investigation.
